# Traitement de la base Sirene
Projet d'identification des déserts médicaux

In [1]:
import os

import pandas as pd

In [ ]:
# Ne pas oublier PEP 8 (cf. pep8_cheatsheet.py)

In [140]:
# Déclaration des constantes
DATA_DIR = '../data/external/'
PROCESSED_DIR = '../data/processed/'

ENTITIES_FILENAME = 'StockUniteLegale_utf8.zip'
ETAB_FILENAME = 'StockEtablissement_utf8.zip'
NAF_REF_FILENAME = 'naf86.xlsx'
PS_CITIES_FILENAME = 'PS_communes.csv'

ACTIVITY_CODE = '86.'

## Préparation des entités

In [3]:
# Lecture du fichier csv
pd.read_csv(os.path.join(DATA_DIR, ENTITIES_FILENAME), #concaténation du répertoire et du nom de fichier
           compression='zip',
           nrows=1000 #limite la lecture à 1000 lignes
           ).head()

siren statutdiffusionunitelegale unitepurgeeunitelegale  \
0   325175                          O                    NaN   
1  1807254                          O                    NaN   
2  5410220                          O                   True   
3  5410345                          O                   True   
4  5410394                          O                   True   

  datecreationunitelegale sigleunitelegale sexeunitelegale prenom1unitelegale  \
0              2000-09-26              NaN               M            THIERRY   
1              1972-05-01              NaN               M     JACQUES-LUCIEN   
2              1954-12-25              NaN               M            GEORGES   
3                     NaN              NaN               M             MICHEL   
4              1954-12-25              NaN               M             ROBERT   

  prenom2unitelegale prenom3unitelegale prenom4unitelegale  \
0                NaN                NaN                NaN   
1                NaN                NaN                NaN   
2                NaN                NaN                NaN   
3                NaN                NaN                NaN   
4             ALFRED                NaN                NaN   

               ...              denominationunitelegale  \
0              ...                                  NaN   
1              ...                                  NaN   
2              ...                                  NaN   
3              ...                                  NaN   
4              ...                                  NaN   

   denominationusuelle1unitelegale  denominationusuelle2unitelegale  \
0                              NaN                              NaN   
1                              NaN                              NaN   
2                              NaN                              NaN   
3                              NaN                              NaN   
4                              NaN                              NaN   

  denominationusuelle3unitelegale  categoriejuridiqueunitelegale  \
0                             NaN                           1000   
1                             NaN                           1000   
2                             NaN                           1000   
3                             NaN                           1000   
4                             NaN                           1000   

  activiteprincipaleunitelegale  nomenclatureactiviteprincipaleunitelegale  \
0                        32.12Z                                    NAFRev2   
1                        85.59A                                    NAFRev2   
2                         22.02                                        NAP   
3                         79.06                                        NAP   
4                         64.42                                        NAP   

  nicsiegeunitelegale  economiesocialesolidaireunitelegale  \
0                  65                                  NaN   
1                  22                                  NaN   
2                  15                                  NaN   
3                  10                                  NaN   
4                  18                                  NaN   

  caractereemployeurunitelegale  
0                             N  
1                             N  
2                             N  
3                             N  
4                             N  

[5 rows x 33 columns]

In [4]:
pd.read_csv(os.path.join(DATA_DIR, ENTITIES_FILENAME),
           compression='zip',
           nrows=1000 
           ).columns

Index(['siren', 'statutdiffusionunitelegale', 'unitepurgeeunitelegale',
       'datecreationunitelegale', 'sigleunitelegale', 'sexeunitelegale',
       'prenom1unitelegale', 'prenom2unitelegale', 'prenom3unitelegale',
       'prenom4unitelegale', 'prenomusuelunitelegale', 'pseudonymeunitelegale',
       'identifiantassociationunitelegale', 'trancheeffectifsunitelegale',
       'anneeeffectifsunitelegale', 'datederniertraitementunitelegale',
       'nombreperiodesunitelegale', 'categorieentreprise',
       'anneecategorieentreprise', 'datedebut', 'etatadministratifunitelegale',
       'nomunitelegale', 'nomusageunitelegale', 'denominationunitelegale',
       'denominationusuelle1unitelegale', 'denominationusuelle2unitelegale',
       'denominationusuelle3unitelegale', 'categoriejuridiqueunitelegale',
       'activiteprincipaleunitelegale',
       'nomenclatureactiviteprincipaleunitelegale', 'nicsiegeunitelegale',
       'economiesocialesolidaireunitelegale', 'caractereemployeurunitelega

In [7]:
(pd.read_csv(os.path.join(DATA_DIR, ENTITIES_FILENAME), #concaténation du répertoire et du nom de fichier
           compression='zip'
           #nrows=1000 #limite la lecture à 1000 lignes
           )
#.query("activiteprincipaleunitelegale.str.contains('86.')")
['activiteprincipaleunitelegale']
.value_counts()
)

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (2,4,5,6,7,8,9,10,11,12,13,17,21,22,24,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


68.20B    1822169
68.20A     551120
94.99Z     506947
70.22Z     371098
56.10A     216662
43.99C     212566
47.99A     211147
01.11Z     205785
56.10C     187764
70.2C      183793
86.90D     171348
81.10Z     159625
47.89Z     158801
64.20Z     157330
93.12Z     153227
55.3A      145974
86.90E     144949
01.21Z     144772
90.01Z     140701
68.31Z     138745
43.34Z     137700
96.02A     137610
96.09Z     137121
43.21A     134155
74.1J      124889
46.19B     121332
86.21Z     120266
90.03A     119947
69.10Z     116920
01.1A      115284
           ...   
30.40Z          4
43.01           4
33.04           4
44.35           4
65.1F           4
13.1Z           4
90.33           3
12.0Z           3
65.1A           3
04.21           3
15.9Q           3
64.1A           3
13.15           3
39.07           3
75.2C           3
90.26           3
05.21           2
42.01           2
19.10Z          2
15.4E           2
11.02           2
00.40           2
90.31           2
96.0Z           2
32.01     

In [8]:
entities_df = (pd.read_csv(os.path.join(DATA_DIR, ENTITIES_FILENAME), 
                           usecols=['activiteprincipaleunitelegale', 'datecreationunitelegale', 'siren',
                                   'denominationunitelegale', 'nomunitelegale', 'etatadministratifunitelegale'
                                   ],
                           compression='zip',
                           #nrows=100000 
                           )
             ).dropna(subset=['activiteprincipaleunitelegale']).query('etatadministratifunitelegale == "A"')

#Filtrage des professionels de santé
#entities_df = entities_df[entities_df.activiteprincipaleunitelegale.str.startswith(ACTIVITY_CODE)]

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [9]:
entities_df.shape #nombre de lignes et de colonnes

(11696143, 6)

In [10]:
entities_df.denominationunitelegale.isnull().sum()

4403344

In [11]:
entities_df.nomunitelegale.isnull().sum()

7292817

In [12]:
entities_df.head()

siren datecreationunitelegale etatadministratifunitelegale  \
0    325175              2000-09-26                            A   
31  5411582              1954-01-01                            A   
38  5420021              1954-01-01                            A   
45  5420120              1954-01-01                            A   
51  5420260              1954-12-25                            A   

   nomunitelegale                denominationunitelegale  \
0         JANOYER                                    NaN   
31       GERVOISE                                    NaN   
38            NaN           ETABLISSEMENTS LUCIEN BIQUEZ   
45            NaN  SOCIETE DES SUCRERIES DU MARQUENTERRE   
51            NaN                SOCIETE ANONYME MILBLED   

   activiteprincipaleunitelegale  
0                         32.12Z  
31                        68.20B  
38                        46.69B  
45                        70.10Z  
51                         20.01

## Préparation des établissements

In [13]:
pd.read_csv(os.path.join(DATA_DIR, ETAB_FILENAME),
           compression='zip',
           nrows=1000 
           ).head()

siren  nic        siret statutdiffusionetablissement  \
0  325175   16  32517500016                            O   
1  325175   24  32517500024                            O   
2  325175   32  32517500032                            O   
3  325175   40  32517500040                            O   
4  325175   57  32517500057                            O   

  datecreationetablissement trancheeffectifsetablissement  \
0                2000-09-26                           NaN   
1                2008-05-20                           NaN   
2                2009-05-27                           NaN   
3                2011-10-21                           NaN   
4                2014-01-07                           NaN   

   anneeeffectifsetablissement activiteprincipaleregistremetiersetablissement  \
0                          NaN                                         3212ZZ   
1                          NaN                                            NaN   
2                          NaN                                            NaN   
3                          NaN                                         3212ZZ   
4                          NaN                                            NaN   

  datederniertraitementetablissement  etablissementsiege  \
0                2015-03-18T00:58:59               False   
1                2011-12-12T09:40:04               False   
2                2014-07-08T00:10:21               False   
3                2015-03-18T00:58:59               False   
4                2018-07-10T14:17:15               False   

                ...                libellepaysetranger2etablissement  \
0               ...                                              NaN   
1               ...                                              NaN   
2               ...                                              NaN   
3               ...                                              NaN   
4               ...                                              NaN   

    datedebut  etatadministratifetablissement   enseigne1etablissement  \
0  2009-05-27                               F                      NaN   
1  2011-10-21                               F  TAHITI PERLES CREATIONS   
2  2011-10-21                               F  TAHITI PERLES CREATIONS   
3  2014-01-07                               F   TAHITI PERLES CREATION   
4  2018-02-07                               F   TAHITI PERLES CREATION   

  enseigne2etablissement enseigne3etablissement  \
0                    NaN                    NaN   
1                    NaN                    NaN   
2                    NaN                    NaN   
3                    NaN                    NaN   
4                    NaN                    NaN   

   denominationusuelleetablissement activiteprincipaleetablissement  \
0                               NaN                          32.12Z   
1                               NaN                          47.89Z   
2                               NaN                          32.12Z   
3                               NaN                          32.12Z   
4                               NaN                          32.12Z   

   nomenclatureactiviteprincipaleetablissement caractereemployeuretablissement  
0                                      NAFRev2                               N  
1                                      NAFRev2                               N  
2                                      NAFRev2                               N  
3                                      NAFRev2                               N  
4                                      NAFRev2                               N  

[5 rows x 48 columns]

In [14]:
pd.read_csv(os.path.join(DATA_DIR, ETAB_FILENAME),
           compression='zip',
           nrows=1000 
           ).columns

Index(['siren', 'nic', 'siret', 'statutdiffusionetablissement',
       'datecreationetablissement', 'trancheeffectifsetablissement',
       'anneeeffectifsetablissement',
       'activiteprincipaleregistremetiersetablissement',
       'datederniertraitementetablissement', 'etablissementsiege',
       'nombreperiodesetablissement', 'complementadresseetablissement',
       'numerovoieetablissement', 'indicerepetitionetablissement',
       'typevoieetablissement', 'libellevoieetablissement',
       'codepostaletablissement', 'libellecommuneetablissement',
       'libellecommuneetrangeretablissement',
       'distributionspecialeetablissement', 'codecommuneetablissement',
       'codecedexetablissement', 'libellecedexetablissement',
       'codepaysetrangeretablissement', 'libellepaysetrangeretablissement',
       'complementadresse2etablissement', 'numerovoie2etablissement',
       'indicerepetition2etablissement', 'typevoie2etablissement',
       'libellevoie2etablissement', 'codepostal2

In [15]:
etab_df = (pd.read_csv(os.path.join(DATA_DIR, ETAB_FILENAME), 
                           usecols=['activiteprincipaleetablissement', 'datecreationetablissement', 'siren',
                                   'denominationusuelleetablissement', 'etatadministratifetablissement', 'enseigne1etablissement',
                                   'codepostaletablissement', 'codecommuneetablissement', 'libellecommuneetablissement'
                                   ],
                           compression='zip',
                           #nrows=100000 
                           )
             ).dropna(subset=['activiteprincipaleetablissement']).query('etatadministratifetablissement == "A"')

#Filtrage des professionels de santé
#etab_df = etab_df[etab_df.activiteprincipaleetablissement.str.startswith(ACTIVITY_CODE)]

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (20,41,44) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [75]:
etab_df.shape

(11233968, 9)

In [59]:
etab_df.head(40)

siren datecreationetablissement  codepostaletablissement  \
892      5720164                1983-09-28                  80100.0   
6424     7050040                1970-01-01                   4510.0   
17575   17150533                1971-01-01                  21000.0   
26233   24080749                2002-12-19                  97600.0   
34605   35204429                2009-01-01                  31200.0   
37513   36220150                1997-01-27                   6600.0   
41402   37240348                1985-01-01                  39100.0   
41851   37739802                2004-10-01                  64100.0   
112251  46620241                1966-01-01                  20117.0   
112800  46750147                1967-01-01                  59500.0   
112801  46750147                2007-11-01                  59187.0   
117335  54390315                2004-01-19                  97600.0   
119320  54805577                1900-01-01                  13011.0   
121539  55806673                1955-01-01                  13011.0   
121603  55807838                2001-10-01                  13008.0   
123474  56803117                1956-01-01                  13400.0   
127556  57818460                1957-01-01                  13006.0   
127558  57818460                2003-07-31                  13008.0   
128666  58504390                2016-06-03                  38180.0   
138059  61800090                1961-01-01                  13009.0   
138689  62200241                2000-07-10                  49800.0   
138690  62200241                2003-07-01                  49800.0   
142127  63803654                2012-03-21                  13290.0   
142161  63804033                1963-01-01                  13600.0   
142182  63804447                1963-01-01                  13004.0   
142615  64200728                1998-04-01                  49300.0   
144945  65201220                2015-02-16                  49290.0   
146440  65804064                1993-05-18                  13011.0   
146536  65805384                1965-01-01                  13002.0   
146549  65805590                1965-01-01                  13600.0   
146992  66200056                2006-08-30                  49800.0   
146993  66200056                2015-10-12                  49800.0   
149744  66803677                1994-09-01                  13012.0   
150243  67200790                1967-01-01                  49370.0   
150777  67501650                2006-06-27                  38130.0   
151567  67800177                1967-01-01                  13600.0   
152087  67801860                1995-01-03                  13400.0   
152170  67803452                1967-01-01                  13400.0   
152171  67803452                2018-10-01                  13008.0   
154742  68802826                1968-01-01                  13190.0   

       libellecommuneetablissement codecommuneetablissement  \
892                      ABBEVILLE                    80001   
6424                        AIGLUN                    04001   
17575                        DIJON                    21231   
26233                    MAMOUDZOU                    97611   
34605                     TOULOUSE                    31555   
37513                      ANTIBES                    06004   
41402                         DOLE                    39198   
41851                      BAYONNE                    64102   
112251                       OCANA                    2A181   
112800                       DOUAI                    59178   
112801                       DECHY                    59170   
117335                   MAMOUDZOU                    97611   
119320                MARSEILLE 11                    13211   
121539                MARSEILLE 11                    13211   
121603                 MARSEILLE 8                    13208   
123474                     AUBAGNE                    13005   
127556                 MARSE

In [60]:
etab_df.dtypes

siren                                 int64
datecreationetablissement            object
codepostaletablissement             float64
libellecommuneetablissement          object
codecommuneetablissement             object
etatadministratifetablissement       object
enseigne1etablissement               object
denominationusuelleetablissement     object
activiteprincipaleetablissement      object
dtype: object

In [61]:
entities_df.dtypes

siren                             int64
datecreationunitelegale          object
etatadministratifunitelegale     object
nomunitelegale                   object
denominationunitelegale          object
activiteprincipaleunitelegale    object
dtype: object

In [64]:
process_df = etab_df.merge(entities_df, on='siren', how='left')

In [65]:
process_df.shape

(560103, 14)

In [67]:
process_df[process_df.activiteprincipaleunitelegale.isnull()].shape

(7954, 14)

In [73]:
etab_df.merge(entities_df, on='siren', how='outer', indicator=True)['_merge'].value_counts()

both          552149
right_only      8994
left_only       7954
Name: _merge, dtype: int64

In [ ]:
#8994 entreprises pour lesquelles on a pas d'établissement (du moins dans la section 86: santé)
# 7954 établissements pour lesquels on a pas d'entité

In [80]:
entities_df.memory_usage(deep=True).sum()/(1024*1024) # volume en mégaoctet

3464.3372735977173

In [81]:
etab_df.merge(entities_df, on='siren', how='outer', indicator=True)['_merge'].value_counts() #merge sans le filtre santé (section 86)

both          11233945
right_only     1622950
left_only           23
Name: _merge, dtype: int64

### faire un git commit dans le git bash:
    git add .
    git commit -m "préparation siren"
    git push -u origin master 

In [25]:
#Base de données mergée avec etab et entities
merge_df = etab_df.merge(entities_df, on='siren', how='outer', indicator=True)

In [26]:
merge_df.head()

siren datecreationetablissement  codepostaletablissement  \
0   325175                2018-02-07                  13004.0   
1  5420021                2009-12-23                  80100.0   
2  5420120                1989-01-27                  80120.0   
3  5420120                1900-01-01                  62180.0   
4  5420120                1900-01-01                  62140.0   

  libellecommuneetablissement codecommuneetablissement  \
0                 MARSEILLE 4                    13204   
1                   ABBEVILLE                    80001   
2                         RUE                    80688   
3              RANG DU FLIERS                    62688   
4                 MARCONNELLE                    62550   

  etatadministratifetablissement enseigne1etablissement  \
0                              A                    NaN   
1                              A                    NaN   
2                              A                    NaN   
3                              A                    NaN   
4                              A                    NaN   

  denominationusuelleetablissement activiteprincipaleetablissement  \
0                              NaN                          32.12Z   
1                              NaN                          46.69B   
2                              NaN                          10.81Z   
3                              NaN                          52.10B   
4                              NaN                          70.10Z   

  datecreationunitelegale etatadministratifunitelegale nomunitelegale  \
0              2000-09-26                            A        JANOYER   
1              1954-01-01                            A            NaN   
2              1954-01-01                            A            NaN   
3              1954-01-01                            A            NaN   
4              1954-01-01                            A            NaN   

                 denominationunitelegale activiteprincipaleunitelegale _merge  
0                                    NaN                        32.12Z   both  
1           ETABLISSEMENTS LUCIEN BIQUEZ                        46.69B   both  
2  SOCIETE DES SUCRERIES DU MARQUENTERRE                        70.10Z   both  
3  SOCIETE DES SUCRERIES DU MARQUENTERRE                        70.10Z   both  
4  SOCIETE DES SUCRERIES DU MARQUENTERRE                        70.10Z   both

In [60]:
#Filtrer les etab en 86 ou les entités en 86 (Filtrage des professionels de santé)
merged_df = merge_df[merge_df.activiteprincipaleetablissement.str.startswith(ACTIVITY_CODE) | merge_df.activiteprincipaleunitelegale.str.startswith(ACTIVITY_CODE)]

In [61]:
merged_df.head()

siren datecreationetablissement  codepostaletablissement  \
58      5720164                1983-09-28                  80100.0   
416     7050040                1970-01-01                   4510.0   
5115   17150533                1971-01-01                  21000.0   
12308  24080749                2002-12-19                  97600.0   
14190  35204429                2009-01-01                  31200.0   

      libellecommuneetablissement codecommuneetablissement  \
58                      ABBEVILLE                    80001   
416                        AIGLUN                    04001   
5115                        DIJON                    21231   
12308                   MAMOUDZOU                    97611   
14190                    TOULOUSE                    31555   

      etatadministratifetablissement enseigne1etablissement  \
58                                 A                    NaN   
416                                A                    NaN   
5115                               A                    NaN   
12308                              A                    NaN   
14190                              A                    NaN   

      denominationusuelleetablissement activiteprincipaleetablissement  \
58                                 NaN                          86.10Z   
416                                NaN                          86.10Z   
5115                               NaN                          86.10Z   
12308                              NaN                          86.90B   
14190                              NaN                          86.23Z   

      datecreationunitelegale etatadministratifunitelegale nomunitelegale  \
58                 1957-01-01                            A            NaN   
416                1970-01-01                            A            NaN   
5115               1971-01-01                            A            NaN   
12308              2002-12-19                            A            NaN   
14190              1983-10-21                            A            NaN   

           denominationunitelegale activiteprincipaleunitelegale _merge  
58              SA SAINTE-ISABELLE                        86.10Z   both  
416              CENTRE DES CARMES                        86.10Z   both  
5115          CLINIQUE LES ROSIERS                        86.10Z   both  
12308              MAYO BIO SELARL                        86.90B   both  
14190  SAURAT PHILIPPE ET ISABELLE                        86.23Z   both

In [62]:
merged_df['_merge'].value_counts()

both          581698
right_only         0
left_only          0
Name: _merge, dtype: int64

In [63]:
#Analyser quels sont les types d'écart entre les activités etab et les entités
#Tableau croisé activité etab et entité
pd.crosstab(merged_df.activiteprincipaleunitelegale.str.startswith(ACTIVITY_CODE), merged_df.activiteprincipaleetablissement.str.startswith(ACTIVITY_CODE)) 

activiteprincipaleetablissement  False   True 
activiteprincipaleunitelegale                 
False                                0    7954
True                             21595  552149

Donc nous avons 7954 unité légales classées comme professionels de santé (86) alors que les établissements correspondants ne le sont pas.

Nous avons également 21595 établissements classés comme professionels de santé (86) alors que les unités légales correspondantes ne le sont pas.

Nous avons 552149 correspondances (établissements et unités légales tous deux classés comme professionels de santé: 86).

In [131]:
ps_df = (merged_df.loc[merged_df.activiteprincipaleunitelegale.str.startswith("86.") |
                      merged_df.activiteprincipaleetablissement.str.startswith("86.")]
         .assign(etab_ape_l1 = lambda df: df.activiteprincipaleetablissement.str.slice(0, 2))
         .assign(el_ape_l1 = lambda x: x.activiteprincipaleunitelegale.str.slice(0, 2))
         .assign(cross_l1 = lambda df: df.el_ape_l1 + '_' + df.etab_ape_l1)
        )

In [73]:
ps_df.cross_l1.value_counts()

86_86    552149
86_68     10368
84_86      3477
86_87      2259
86_88      1448
86_70      1405
86_01      1336
88_86       944
65_86       772
86_47       761
86_85       733
86_66       617
86_96       408
86_90       357
87_86       352
86_55       342
01_86       332
68_86       313
86_49       302
47_86       257
86_02       242
85_86       214
96_86       207
94_86       205
86_35       180
70_86       149
90_86       116
86_74        89
86_82        87
49_86        85
          ...  
86_13         3
71_86         2
86_38         2
33_86         2
86_00         2
86_65         2
86_15         2
13_86         2
86_91         2
41_86         2
79_86         2
21_86         1
86_27         1
86_08         1
86_11         1
28_86         1
50_86         1
86_80         1
11_86         1
78_86         1
86_22         1
86_61         1
31_86         1
86_25         1
15_86         1
05_86         1
86_29         1
86_79         1
86_18         1
14_86         1
Name: cross_l1, Length: 

Donc on a juste besoin des 4 premières catégories pour avoir la majorité de la population.

In [70]:
#Dans anaconda prompt: pip install dpaste_magic
%load_ext dpaste_magic

The dpaste_magic extension is already loaded. To reload it, use:
  %reload_ext dpaste_magic


In [ ]:
#Pour copier-coller automatiquement le code de quelqu'un d'autre:
#%getdpaste SWEe

In [76]:
# Création d'une cross table avec une seule ligne
# Remplacer "process_df" par "merged_df" ou le nom choisi pour désigner le DF créé à partir de l'outer join.

pd.crosstab((ps_df[ps_df.activiteprincipaleetablissement.str.startswith(ACTIVITY_CODE)]).activiteprincipaleunitelegale,
            (ps_df[ps_df.activiteprincipaleunitelegale.str.startswith(ACTIVITY_CODE)]).activiteprincipaleetablissement
           )

activiteprincipaleetablissement  86.10Z  86.21Z  86.22A  86.22B  86.22C  \
activiteprincipaleunitelegale                                             
86.10Z                             8736     312      11       4      17   
86.21Z                               12   88879     207      97     634   
86.22A                                8     202    7618       0     216   
86.22B                                2      68       0    6410     186   
86.22C                                1     423     100     106   47899   
86.23Z                                0      15       0       6      12   
86.90A                                0       0       0       0       0   
86.90B                                0       6       0       0      10   
86.90C                                0       6       0       0       0   
86.90D                                6      19       1       0       1   
86.90E                                1       9       0       0       2   
86.90F                                0      10       0       1       8   

activiteprincipaleetablissement  86.23Z  86.90A  86.90B  86.90C  86.90D  \
activiteprincipaleunitelegale                                             
86.10Z                                9      35      15      14     322   
86.21Z                                4       0      10       0      15   
86.22A                                0       0       3       0       2   
86.22B                               16       0       0       0       1   
86.22C                                7       0      24       0       5   
86.23Z                            44336       0       0       0       4   
86.90A                                0    7866       0       0       0   
86.90B                                0       0    6450       0       3   
86.90C                                0       1       1     422       0   
86.90D                                1       1       3       0  127363   
86.90E                                1       0       0       0      85   
86.90F                                0       0       0       0      61   

activiteprincipaleetablissement  86.90E  86.90F  
activiteprincipaleunitelegale                    
86.10Z                                4       9  
86.21Z                               12      22  
86.22A                                0       1  
86.22B                                0       0  
86.22C                                2       9  
86.23Z                                1       2  
86.90A                                0       3  
86.90B                                1       8  
86.90C                                0       0  
86.90D                              158     251  
86.90E                           128853     169  
86.90F                              349   72954

In [77]:
pd.crosstab(ps_df.query('cross_l1 == "86_86"').activiteprincipaleetablissement,
            ps_df.query('cross_l1 == "86_86"').activiteprincipaleunitelegale)

activiteprincipaleunitelegale    86.10Z  86.21Z  86.22A  86.22B  86.22C  \
activiteprincipaleetablissement                                           
86.10Z                             8736      12       8       2       1   
86.21Z                              312   88879     202      68     423   
86.22A                               11     207    7618       0     100   
86.22B                                4      97       0    6410     106   
86.22C                               17     634     216     186   47899   
86.23Z                                9       4       0      16       7   
86.90A                               35       0       0       0       0   
86.90B                               15      10       3       0      24   
86.90C                               14       0       0       0       0   
86.90D                              322      15       2       1       5   
86.90E                                4      12       0       0       2   
86.90F                                9      22       1       0       9   

activiteprincipaleunitelegale    86.23Z  86.90A  86.90B  86.90C  86.90D  \
activiteprincipaleetablissement                                           
86.10Z                                0       0       0       0       6   
86.21Z                               15       0       6       6      19   
86.22A                                0       0       0       0       1   
86.22B                                6       0       0       0       0   
86.22C                               12       0      10       0       1   
86.23Z                            44336       0       0       0       1   
86.90A                                0    7866       0       1       1   
86.90B                                0       0    6450       1       3   
86.90C                                0       0       0     422       0   
86.90D                                4       0       3       0  127363   
86.90E                                1       0       1       0     158   
86.90F                                2       3       8       0     251   

activiteprincipaleunitelegale    86.90E  86.90F  
activiteprincipaleetablissement                  
86.10Z                                1       0  
86.21Z                                9      10  
86.22A                                0       0  
86.22B                                0       1  
86.22C                                2       8  
86.23Z                                1       0  
86.90A                                0       0  
86.90B                                0       0  
86.90C                                0       0  
86.90D                               85      61  
86.90E                           128853     349  
86.90F                              169   72954

In [95]:
#pour faire apparaître la table qu'on a copié-collé
naf_df = pd.read_clipboard()

In [96]:
naf_df

NAF                                            Libellé  \
0   86.10Z                            Activités hospitalières   
1   86.21Z                 Activité des médecins généralistes   
2   86.22A   Activités de radiodiagnostic et de radiothérapie   
3   86.22B                            Activités chirurgicales   
4   86.22C         Autres activités des médecins spécialistes   
5   86.23Z                                  Pratique dentaire   
6   86.90A                                         Ambulances   
7   86.90B                  Laboratoires d'analyses médicales   
8   86.90C           Centres de collecte et banques d'organes   
9   86.90D       Activités des infirmiers et des sages-femmes   
10  86.90E  Activités des professionnels de la rééducation...   
11  86.90F   Activités de santé humaine non classées ailleurs   

                Catégorie  
0                 hopital  
1             generaliste  
2            radiologiste  
3              chirurgien  
4       autre_specialiste  
5                dentiste  
6               ambulance  
7        analyse_medicale  
8                  organe  
9              infirmiers  
10  reeducateur_podologue  
11                  autre

In [97]:
#pour enregistrer la table qu'on vient de faire apparaître au format excel
naf_df.to_excel(os.path.join(DATA_DIR, NAF_REF_FILENAME), index = False)

## Préparation  de la table des codes NAF

In [101]:
#coder le code d'activité 86 (règle de choix): 86_86 -> etab 86_x ou y -> 86
ps_df.loc[ps_df.cross_l1 == "86_86", 'code_activite'] = ps_df.activiteprincipaleetablissement
ps_df.loc[ps_df.el_ape_l1 != "86", 'code_activite'] = ps_df.activiteprincipaleetablissement
ps_df.loc[~ps_df.activiteprincipaleetablissement.str.startswith(ACTIVITY_CODE), 'code_activite'] = ps_df.activiteprincipaleunitelegale

In [103]:
#autre manière de faire:
ps_df['code_activite'] = ps_df.apply(lambda x: x.activiteprincipaleetablissement 
                                     if x.etab_ape_l1 == "86"
                                     else x.activiteprincipaleunitelegale, 
                                     axis=1
                                     )

In [ ]:
#troisème version:
ps_df['code_activite'] = ps_df.activiteprincipaleetablissement
ps_df['code_activite'] = ps_df.where(ps_df.etab_ape_l1 != "86", 
                                     ps_df.activiteprincipaleunitelegale,
                                     axis= 0
                                     )

In [134]:
#quatrième version:
import numpy as np

ps_df['code_activite'] = np.where(ps_df.etab_ape_l1 == '86',
                                  ps_df.activiteprincipaleetablissement,
                                  ps_df.activiteprincipaleunitelegale
                                  )

In [135]:
ps_df.code_activite.value_counts()

86.90E    132626
86.90D    131850
86.21Z     97022
86.90F     75981
86.22C     51576
86.23Z     47039
86.10Z     14087
86.90A      8687
86.22A      8508
86.22B      7103
86.90B      6779
86.90C       440
Name: code_activite, dtype: int64

In [136]:
ps_df.columns

Index(['siren', 'datecreationetablissement', 'codepostaletablissement',
       'libellecommuneetablissement', 'codecommuneetablissement',
       'etatadministratifetablissement', 'enseigne1etablissement',
       'denominationusuelleetablissement', 'activiteprincipaleetablissement',
       'datecreationunitelegale', 'etatadministratifunitelegale',
       'nomunitelegale', 'denominationunitelegale',
       'activiteprincipaleunitelegale', '_merge', 'naf', 'etab_ape_l1',
       'el_ape_l1', 'cross_l1', 'code_activite'],
      dtype='object')

## Ajout du type de PS et sauvegarde

In [141]:
#Base de données mergée avec naf, puis tranposée, puis enregistrée en csv
(ps_df
 .merge(naf_df, how='inner', left_on='code_activite', right_on='NAF')
 [['codecommuneetablissement', 'Catégorie']]
 .dropna(subset=['codecommuneetablissement'])
 .assign(count = 1)
 .groupby(['codecommuneetablissement', 'Catégorie'])
 .sum()
 .reset_index()
 .pivot(index='codecommuneetablissement', columns='Catégorie', values='count')
 .fillna(0)
).to_csv(os.path.join(PROCESSED_DIR, PS_CITIES_FILENAME), sep=';')

### Dans le git bash:

cd deserts_medicaux

cd deserts_medicaux

git status (pour voir ce qui est en cours dans le répo)

git commit -m "finaliser preparation sirene"